### PR TITLE
Adds load times to launchpad telemetry context, and adds event for slow loads

### PR DESCRIPTION
**Pull request by axosoft-ramint**

### CODE REVIEW

**Comment by axosoft-ramint** [comment]
Some explanation for the current implementation not simply putting the load times on the "on Refreshed" event of the Focus Provider:
1. Having quickpick commands use event listeners feels a bit weird, and I couldn't find an example of any such commands that set up a disposable and listen for events from other services.
2. Because of 1, my current implementation plumbs the load times through the functions that fetch the focus items, which feels weird and causes all these updates to shapes to account for those load times.

If there's a way we can get away with returning just the items on `getCategorizedItems` and returning the load times separately, I'd love to hear some feedback on that.

**Comment by eamodio** [comment]
Hey @axosoft-ramint, while reviewing your PR, I'd suggest the following code changes:

👉 [**Used new util functions**](https://gitkraken.dev/drafts/efefdae8-7064-43c7-8cbc-74d96fdd45ec?source=PRComment)
>Added new util functions (already on `main`), and it should now send the slow event anytime it crosses the threshold, but will now capture the full duration even it if errors out

You can also review and apply these suggestions [locally on your machine](https://gitkraken.dev/link/drafts/efefdae8-7064-43c7-8cbc-74d96fdd45ec?type=suggested_pr_change&prEntityId=WyJnaXRodWIiLCJwciIsIjEiLCIiLCJQUl9rd0RPQS1NUnVNNXhWaWwzIl0=&source=PRComment).

<details><summary>Learn more about GitKraken Code Suggest</summary>

>[Code Suggest](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbS9zb2x1dGlvbnMvY29kZS1zdWdnZXN0P3NvdXJjZT1naXRodWIuY29tJm1lZGl1bT1wcl9jb21tZW50?source=PRSuggest&event=redirectToCodeSuggestPage&provider=github) liberates your code reviews from GitHub's restrictive, comment-only feedback style. As simple as suggesting changes in a Google-doc, provide real code suggestions from where you code, e.g. your IDE, and on anything in your project — not just on the lines of code changed in the PR.

_Join your team on [GitKraken](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbT9zb3VyY2U9Z2l0aHViLmNvbSZtZWRpdW09cHJfY29tbWVudA==?source=PRSuggest&event=redirectToIndexPage&provider=github) to speed up PR review._
</details>

**Review by eamodio — CHANGES_REQUESTED** [review]
Changes in attached Code Suggest

File: src/constants.ts [hunk 1]
```diff
@@ -1344,6 +1344,8 @@ export type TelemetryEvents = {
 	'launchpad/indicator/hidden': TelemetryEventData;
 	/** Sent when the launchpad indicator loads (with data) for the first time ever for this device */
 	'launchpad/indicator/firstLoad': TelemetryEventData;
+	/** Sent when a launchpad operation is taking longer than a set timeout to complete */
+	'launchpad/operation/slow': TelemetryEventData;
```
Comment (eamodio) [inline src/constants.ts:1348]: Please fill in the specific typed payload here

Comment (axosoft-ramint) [reply src/constants.ts]: Done in commit: 665532eaf4fef604c72dc1cdeb4a9691fa93bda4

File: src/plus/focus/focusProvider.ts [hunk 2]
```diff
@@ -836,3 +877,18 @@ function ensureRemoteUrl(url: string) {
...
+
+function withDurationAndSlowEventOnTimeout<T>(
+	promise: Promise<T>,
+	name: 'getMyPullRequests' | 'getCodeSuggestionCounts' | 'getCodeSuggestions' | 'getEnrichedItems',
+	container: Container,
+): Promise<TimedResult<T>> {
+	return timedWithSlowThreshold(promise, {
+		timeout: slowEventTimeout,
+		onSlow: (_duration: number) => {
+			container.telemetry.sendEvent('launchpad/operation/slow', { timeout: slowEventTimeout, operation: name });
```
Comment (eamodio) [inline src/plus/focus/focusProvider.ts:891]: Definitely want to capture the duration here in this event

File: src/plus/focus/focusProvider.ts [hunk 3]
```diff
@@ -173,6 +174,13 @@ export type FocusRefreshEvent =
 
 export const supportedFocusIntegrations = [HostingIntegrationId.GitHub];
 
+export interface FocusItemsWithDurations {
+	items: FocusItem[];
+	prsDuration: number | undefined;
```
Comment (eamodio) [inline src/plus/focus/focusProvider.ts:179]: I'd suggest to move these durations into a `timings` or `durations` object

Comment (eamodio) [reply src/plus/focus/focusProvider.ts:179]: Probably would suggest `timings` as both the prop and the type name

File: src/plus/focus/focus.ts [hunk 4]
```diff
@@ -1013,12 +1030,15 @@ async function updateContextItems(container: Container, context: Context, option
...
 
-	const grouped = countFocusItemGroups(context.items);
+	const grouped = countFocusItemGroups(context.items.items);
 
 	const updatedContext: NonNullable<(typeof context)['telemetryContext']> = {
 		...context.telemetryContext,
-		'items.count': context.items.length,
+		'items.count': context.items.items.length,
 		'groups.count': grouped.size,
+		'items.prsDuration': context.items.prsDuration,
```
Comment (eamodio) [inline src/plus/focus/focus.ts:1039]: Would suggest `items.timings.prs`

File: src/plus/focus/focus.ts [hunk 5]
```diff
@@ -1013,12 +1030,15 @@ async function updateContextItems(container: Container, context: Context, option
...
+		'items.codeSuggestionCountsDuration': context.items.codeSuggestionCountsDuration,
```
Comment (eamodio) [inline src/plus/focus/focus.ts:1040]: Would suggest `items.timings.codeSuggestionCounts`

File: src/plus/focus/focus.ts [hunk 6]
```diff
@@ -1013,12 +1030,15 @@ async function updateContextItems(container: Container, context: Context, option
...
+		'items.enrichedItemsDuration': context.items.enrichedItemsDuration,
```
Comment (eamodio) [inline src/plus/focus/focus.ts:1041]: Would suggest `items.timings.enrichedItems`

**Comment by eamodio** [comment]
After those comments above are resolved, :shipit:

**Review by eamodio — APPROVED** [review]
After those comments above are resolved, :shipit:

**Pull request merged**